aws/karpenter-core
# BREAKING CHANGE: replace node.kubernetes.io/unschedulable with a Karpenter-specific taint

Fixes #N/A 

**Description**
Design Doc: https://github.com/aws/karpenter-core/pull/585

Karpenter will now use the `karpenter.sh/disrupting` taint instead of `node.kubernetes.io/unschedulable` taint when deprovisioning nodes. Karpenter only adds this taint once it has begun spinning up replacement nodes for deprovisioning actions, and subsequently deleting the candidate node. 

- Karpenter would add/remove this taint, assuming no other agent had been managing this taint.
- This fixes an issue where in rare cases where Karpenter restarted while executing a deprovisioning action, Karpenter can leave nodes that were meant to be deprovisioned as cordoned. 
- This breaks some users with pods that tolerate the `node.kubernetes.io/unschedulable` taint or tolerate all keys: a `*` toleration. If users want to ensure their daemonsets and other pods that had this toleration are not evicted as part of termination, they'll need to add this toleration to their pods. 
- This also modifies the cluster state logic to remove the `stateNode.MarkedForDeletion` variable. Karpenter can now rely on the existence of the `karpenter.sh/disrupting` taint to know when a node has been marked for deletion.

**How was this change tested?**
- make presubmit 
- make apply && manual test

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
